src/transformers/generation_utils.py

```diff
@@ -1882,28 +1882,34 @@ def contrastive_search(
             model_inputs = self.prepare_inputs_for_generation(input_ids, **model_kwargs)
 
             # if the first step in the loop, encode all the prefix and obtain three parameters: (1) past_key_values;
-            # (2) last_hidden_states; (3) logit_for_next_step
+            # (2) last_hidden_states; (3) logit_for_next_step; (4) update model kwargs for the next step
             if model_kwargs.get("past") is None:
                 # encode the given prefix and prepare model inputs; encoder-decoder model process the prefix and save
                 # the `encoder_outputs`
-                output = self(**model_inputs, output_hidden_states=True, output_attentions=True)
+                outputs = self(
+                    **model_inputs, return_dict=True, output_hidden_states=True, output_attentions=output_attentions
+                )
 
                 # past_key_values is required for fast decoding
-                if "past_key_values" not in output:
+                if "past_key_values" not in outputs:
                     raise ValueError(
                         f"{self.__class__.__name__} cannot return `past_key_values` and can therefore **not** be used "
                         "for contrastive search."
                     )
-                past_key_values = output.past_key_values
+                past_key_values = outputs.past_key_values
 
                 # last decoder hidden states will be used to compute the degeneration penalty (cosine similarity with
                 # previous tokens)
                 if self.config.is_encoder_decoder:
-                    last_hidden_states = output.decoder_hidden_states[-1]
+                    last_hidden_states = outputs.decoder_hidden_states[-1]
                 else:
-                    last_hidden_states = output.hidden_states[-1]
+                    last_hidden_states = outputs.hidden_states[-1]
                 # next logit for contrastive search to select top-k candidate tokens
-                logit_for_next_step = output.logits[:, -1, :]
+                logit_for_next_step = outputs.logits[:, -1, :]
+
+                model_kwargs = self._update_model_kwargs_for_generation(
+                    outputs, model_kwargs, is_encoder_decoder=self.config.is_encoder_decoder
+                )
```

Comment:
The most important change: updating the attention mask was happening manually below, and we already had a function that took care of it. While updating the mask is simple in PT, in TF with XLA this amounts to >100 lines of code.